openedx/core/djangoapps/content_libraries/api/container_metadata.py

```diff
@@ -0,0 +1,145 @@
+        except StopIteration:
+            raise ValueError(f"no container_type for XML tag: <{olx_tag}>") from None
+
+
+@dataclass(frozen=True, kw_only=True)
+class ContainerMetadata(PublishableItem):
+    """
+    Class that represents the metadata about a Container (e.g. Unit) in a content library.
+    """
+    container_key: LibraryContainerLocator
+    container_type: ContainerType
+    container_pk: int
+
+    @classmethod
+    def from_container(cls, library_key, container: Container, associated_collections=None):
+        """
+        Construct a ContainerMetadata object from a Container object.
+        """
+        last_publish_log = container.versioning.last_publish_log
+        container_key = library_container_locator(
+            library_key,
+            container=container,
+        )
+        container_type = ContainerType(container_key.container_type)
+        published_by = ""
+        if last_publish_log and last_publish_log.published_by:
+            published_by = last_publish_log.published_by.username
+
+        draft = container.versioning.draft
+        published = container.versioning.published
+        last_draft_created = draft.created if draft else None
+        if draft and draft.publishable_entity_version.created_by:
+            last_draft_created_by = draft.publishable_entity_version.created_by.username
+        else:
+            last_draft_created_by = ""
+        tags = get_object_tag_counts(str(container_key), count_implicit=True)
+
+        return cls(
+            container_key=container_key,
+            container_type=container_type,
+            container_pk=container.pk,
+            display_name=draft.title,
+            created=container.created,
+            modified=draft.created,
+            draft_version_num=draft.version_num,
+            published_version_num=published.version_num if published else None,
+            published_display_name=published.title if published else None,
+            last_published=None if last_publish_log is None else last_publish_log.published_at,
+            published_by=published_by,
+            last_draft_created=last_draft_created,
+            last_draft_created_by=last_draft_created_by,
+            has_unpublished_changes=authoring_api.contains_unpublished_changes(container.pk),
+            tags_count=tags.get(str(container_key), 0),
+            collections=associated_collections or [],
+        )
+
+
+def library_container_locator(
+    library_key: LibraryLocatorV2,
+    container: Container,
+) -> LibraryContainerLocator:
+    """
+    Returns a LibraryContainerLocator for the given library + container.
+    """
+    container_type = None
+    if hasattr(container, 'unit'):
+        container_type = ContainerType.Unit
+    elif hasattr(container, 'subsection'):
+        container_type = ContainerType.Subsection
+    elif hasattr(container, 'section'):
+        container_type = ContainerType.Section
+
+    assert container_type is not None
```

Comment:
```suggestion
    else:
        raise ValueError("Unknown/invalid container type")
```
`assert` doesn't always run in production (if `-O` is used, etc.) so it's probably better to do this.

Reply:
Thanks! Fixed here: efb7de701fbf337c2f32eeac422f9f8270a24b12